vllm/model_executor/layers/mamba/mamba_mixer2.py

```diff
@@ -0,0 +1,494 @@
+        # - IF, world_size divides groups, then sharding
+        #   (n_groups / world_size, n_heads / world_size)
+        #   also maintains the invariant n_heads % n_groups == 0
+        # - HOWEVER IF, world_size DOES NOT divide groups, then we need
+        #   to allocate extra space in the shard, such that groups
+        #   may be replicated to follow the head shard.
+        self.tp_size = get_tensor_model_parallel_world_size()
+        tp_rank = get_tensor_model_parallel_rank()
+
+        self.ssm_state_size = ssm_state_size
+        self.use_rms_norm = use_rms_norm
+        self.activation = activation
+
+        self.chunk_size = chunk_size
+        self.intermediate_size = intermediate_size
+        self.head_dim = head_dim
+        self.num_heads = num_heads
+
+        self.n_groups = n_groups
+        if n_groups % self.tp_size != 0:
+            # - for TP we shard conv_dim by sharding on n_groups,
+            # - but if n_groups cannot divide tp_size, we need to
+            #   extend some extra groups
+            self.n_groups = n_groups + extra_groups_for_head_shards(
+                n_groups, self.tp_size)
+
+        self.conv_dim = (intermediate_size +
+                         2 * self.n_groups * ssm_state_size)
+        self.conv1d = ColumnParallelLinear(
+            input_size=conv_kernel_size,
+            output_size=self.conv_dim,
+            bias=use_conv_bias,
+            quant_config=None,
+        )
+        # unsqueeze to fit conv1d weights shape into the linear weights shape.
+        # Can't do this in `weight_loader` since it already exists in
+        # `ColumnParallelLinear` and `set_weight_attrs`
+        # doesn't allow to override it
+        self.conv1d.weight.data = self.conv1d.weight.data.unsqueeze(1)
+
+        self.in_proj = ColumnParallelLinear(input_size=hidden_size,
+                                            output_size=intermediate_size +
+                                            self.conv_dim + self.num_heads,
+                                            bias=use_bias,
+                                            quant_config=quant_config)
+
+        # - because in_proj is a concatenation of 3 weights, we
+        #   need to interleave them before sharding
+        # - use the custom weight loader mamba_v2_sharded_weight_loader
+        #   for conv1d.bias, covn1d.weight and in_proj.weight
+        # - need to set these settings, to assign the groups to the head shards
+        group_shard_settings = (
+            self.n_groups * self.ssm_state_size,  # expected model size
+            (self.n_groups - n_groups) *
+            self.ssm_state_size,  # extra dims assigned
+            self.num_heads //
+            n_groups,  # ratio for mapping back to original group
+        )
+        intemediate_settings = (intermediate_size, 0, 1)
+        head_setings = (self.num_heads, 0, 1)
+
+        delattr(self.conv1d.bias, "weight_loader")
```

Comment:
Do we need to do this? I would have thought `set_weight_attrs` was sufficient.

Reply:
yea if I do not have that this will raise because i think the `weight` already has that attribute

```
File [vllm/vllm/model_executor/utils.py:29](vllm/vllm/model_executor/utils.py#line=28), in set_weight_attrs(weight, weight_attrs)
     27     return
     28 for key, value in weight_attrs.items():
---> 29     assert not hasattr(
     30         weight, key), (f"Overwriting existing tensor attribute: {key}")
     32     # NOTE(woosuk): During weight loading, we often do something like:
     33     # narrowed_tensor = param.data.narrow(0, offset, len)
     34     # narrowed_tensor.copy_(real_weight)
   (...)
     39     # we sync the param tensor after its weight loader is called.
     40     # TODO(woosuk): Remove this hack once we have a better solution.
     41     if current_platform.is_tpu() and key == "weight_loader":

AssertionError: Overwriting existing tensor attribute: weight_loader
```

Reply:
I added a comment regarding the need for `delattr` [here](https://github.com/vllm-project/vllm/pull/10909#:~:text=%E2%80%A6-,400db27,-Merge%20state)